### ISSUE TITLE
AndroidRuntimeException: Only the original thread that created a view hierarchy can touch its views.

### ISSUE THREAD

**Issue by semanticer** [toggl/mobileapp#6041]
## 💥 Crash Report
It looks like we are setting visibility on some view outside of the UI thread and based on the events - it looks like the problem connected to `SuggestionsPresented` event. 

### 👥 Impact
0.07% of our android 2.4.1 users (0.1% with the second very similar issue.

### 🔗 Context
?? TODO

### 🛰 Link
https://appcenter.ms/orgs/Toggl/apps/Toggl-Android/crashes/errors/3879922110u/overview
and very similar https://appcenter.ms/orgs/Toggl/apps/Toggl-Android/crashes/errors/195068104u/overview  

### 📝  Exception & Stack Trace
`Android.Util.AndroidRuntimeException: Only the original thread that created a view hierarchy can touch its views.`
```cs
Xamarin Exception Stack:
Android.Util.AndroidRuntimeException: Only the original thread that created a view hierarchy can touch its views.
  at Java.Interop.JniEnvironment+InstanceMethods.CallVoidMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args) [0x00069] in <8cb4399a548b410e90524637e592fa0c>:0
  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeVirtualVoidMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) <0x7d893b0fa4 + 0x0008b> in <8cb4399a548b410e90524637e592fa0c>:0
  at Android.Views.View.set_Visibility (Android.Views.ViewStates value) <0x7d895d6e68 + 0x000af> in <220a4a34f2b349e1b330bc52104afc0e>:0
  at Toggl.Droid.Extensions.Reactive.ViewExtensions+<>c__DisplayClass3_0.<IsVisible>b__0 (System.Boolean isVisible) <0x7d87267bf4 + 0x00063> in <699af8654b494bab8f9e06f420784e0a>:0
  at System.Reactive.AnonymousSafeObserver`1[T].OnNext (T value) [0x0000a] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Sink`1[TTarget].ForwardOnNext (TTarget value) [0x00000] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Linq.ObservableImpl.CombineLatest`3+_+FirstObserver[TFirst,TSecond,TResult].OnNext (TFirst value) [0x00067] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Sink`1[TTarget].ForwardOnNext (TTarget value) [0x00000] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.IdentitySink`1[T].OnNext (T value) [0x00000] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Subjects.BehaviorSubject`1[T].OnNext (T value) [0x0004a] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.AnonymousSafeObserver`1[T].OnNext (T value) [0x0000a] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Sink`1[TTarget].ForwardOnNext (TTarget value) [0x00000] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Linq.ObservableImpl.FirstAsync`1+Sequence+_[TSource].OnNext (TSource value) [0x00007] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Sink`1[TTarget].ForwardOnNext (TTarget value) [0x00000] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.IdentitySink`1[T].OnNext (T value) [0x00000] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Subjects.AsyncSubject`1[T].OnCompleted () [0x00064] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Linq.QueryLanguage+<>c__189`1[TResult].<ToAsync>b__189_1 (System.ValueTuple`2[T1,T2] state) [0x00031] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Concurrency.Scheduler+<>c__75`1[TState].<ScheduleAction>b__75_0 (System.Reactive.Concurrency.IScheduler _, System.ValueTuple`2[T1,T2] tuple) [0x00000] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Concurrency.UserWorkItem`1[TState].Run () [0x0000d] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Concurrency.DefaultScheduler+<>c__5`1[TState].<Schedule>b__5_0 (System.Object closureWorkItem) [0x00000] in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Reactive.Concurrency.ConcurrencyAbstractionLayerImpl+<>c.<QueueUserWorkItem>b__3_0 (System.Object itemObject) <0x7d88b06cc8 + 0x00057> in <8bafe8d5ddd64523922fd4b86ca8b6b5>:0
  at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context (System.Object state) <0x7d8b8a4518 + 0x00053> in <a0731ddcd7084366b5f228dd9713b91f>:0
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) <0x7d8b89eba0 + 0x0018f> in <a0731ddcd7084366b5f228dd9713b91f>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) <0x7d8b89eafc + 0x00047> in <a0731ddcd7084366b5f228dd9713b91f>:0
  at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () <0x7d8b8a44ac + 0x00043> in <a0731ddcd7084366b5f228dd9713b91f>:0
  at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in <a0731ddcd7084366b5f228dd9713b91f>:0
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () <0x7d8b8a43dc + 0x0001f> in <a0731ddcd7084366b5f228dd9713b91f>:0
  at android.view.ViewRootImpl$CalledFromWrongThreadException: Only the original thread that created a view hierarchy can touch its views.
  at android.view.ViewRootImpl.checkThread(ViewRootImpl.java:8191)
  at android.view.ViewRootImpl.requestLayout(ViewRootImpl.java:1420)
  at android.view.View.requestLayout(View.java:24469)
  at android.view.View.requestLayout(View.java:24469)
  at android.view.View.requestLayout(View.java:24469)
  at android.view.View.requestLayout(View.java:24469)
  at android.view.View.requestLayout(View.java:24469)
  at android.view.View.requestLayout(View.java:24469)
  at android.support.constraint.ConstraintLayout.requestLayout(ConstraintLayout.java:3112)
  at android.view.View.requestLayout(View.java:24469)
  at android.view.View.requestLayout(View.java:24469)
  at android.view.View.requestLayout(View.java:24469)
  at android.support.v4.widget.NestedScrollView.requestLayout(NestedScrollView.java:1728)
  at android.view.View.requestLayout(View.java:24469)
  at android.view.View.requestLayout(View.java:24469)
  at android.view.View.setFlags(View.java:15202)
  at android.view.View.setVisibility(View.java:10851)

```

**Comment by semanticer** [comment]
I updated this with the latest impact data and added one very similar and maybe a connected issue. Do you think we can look at this in 2.6 @kornelijepetak ?

**Comment by daividssilverio** [comment]
It looks like 2.6 was the last time this crash happened in production. It seems like this issue has been fixed   in 2.7.X. Sounds like we forgot to add a `AsDriver` somewhere.
